### PR TITLE
Types refactor phase 1

### PR DESCRIPTION
**Pull request by slindley**
This PR addresses #637.

It collapses the representation of types into a single monolithic OCaml type. The end goal is to allow commonality between different kinds of types to be shared (e.g. we should be able to treat type variables, recursive types, and type aliases of any kind uniformly). In the process of switching to a single monolithic OCaml type, the PR also changes the representation of types to use a plain variant type rather than a polymorphic variant type.

The PR does not itself implement the functionality to allow e.g. row aliases to be defined, but it does provide a crucial stepping stone to doing so. Moreover, the current PR maintains various invariants that we will ultimately want to remove - such as the invariant that a `Var` constructor (type variable) or `Recursive` constructor (recursive type) can only appear directly inside a `Meta` constructor (meta variable).

### CODE REVIEW

File: core/instantiate.ml [hunk 1]
```diff
@@ -212,46 +218,35 @@ let instantiate_datatype = instantiates ->- fst3
...
-        let tenv, renv, penv, tys, _qs =
+          let rigidity = if rigid then `Rigid else `Flexible in
+          let new_var = Var (var', kind, rigidity) in
+          let point = Unionfind.fresh new_var in
+          let ty =
+            let open PrimaryKind in
+            match Kind.primary_kind kind with
+            | (Type | Presence) as pk -> pk, Meta point
+            | Row -> Row, Row (StringMap.empty, point, false) in
+          (* let ty = Kind.primary_kind kind, Meta point in *)
```
Comment (jamescheney) [inline core/instantiate.ml:236]: Remove dead code?

File: core/instantiate.ml [hunk 2]
```diff
@@ -333,23 +328,24 @@ let presence = instantiates ->- thd3
...
-             (IntMap.add var t tenv, renv, penv)
-         | (var, (Row, _subkind)), `Row row ->
-             (tenv, IntMap.add var row renv, penv)
-         | (var, (Presence, _subkind)), `Presence f ->
-             (tenv, renv, IntMap.add var f penv)
-         | _ ->
-             raise (internal_error
-               ("Kind mismatch in type application: " ^
+    (fun q ((pk, _) as tyarg) inst_map ->
+      (* TypeUtils.check_type_wellformedness None tyarg; *)
```
Comment (jamescheney) [inline core/instantiate.ml:334]: Remove dead/debugging code?

File: core/instantiate.mli [hunk 3]
```diff
@@ -1,7 +1,8 @@
 open CommonTypes
 open Utility
 
-type instantiation_maps = (Types.datatype IntMap.t * Types.row IntMap.t * Types.field_spec IntMap.t)
+(* TODO: after Types refactoring, turn this into singular *)
```
Comment (jamescheney) [inline core/instantiate.mli:4]: remove "done" comment

File: core/irCheck.ml [hunk 4]
```diff
@@ -164,43 +167,43 @@ end
...
     let lookupVar lvar map =
       match IntMap.find_opt lvar map with
       | Some rvar' -> rvar'
       | None       -> lvar in
-    let handle_variable primary_kind (lid, lsk, lfd) (rid, rsk, rfd) ctx =
+    let handle_variable  (lid, (lpk, lsk), lfd) (rid, (rpk, rsk), rfd) ctx =
       let subst_map, kind_env = ctx.typevar_subst, ctx.tyenv in
       let rvar' = lookupVar lid subst_map in
-      let is_equal = rid = rvar' && lsk = rsk && lfd = rfd in
+      let is_equal = rid = rvar' && lpk = rpk &&lsk = rsk && lfd = rfd in
```
Comment (jamescheney) [inline core/irCheck.ml:199]: insert a space

File: core/irCheck.ml [hunk 5]
```diff
@@ -384,16 +428,19 @@ let eq_types occurrence : type_eq_context -> (Types.datatype * Types.datatype) -
...
+      | Recursive _, _
+      | _, Recursive _ -> Debug.print "IR typechecker encountered recursive type"; true
       | _ ->  false
-    and eq_type_args  (context, l, r)  =
-      match l,r with
-      | `Type lt, `Type rt -> eqt (context, lt, rt)
-      | `Row lr, `Row rr -> eq_rows (context, lr, rr)
-      | `Presence lf, `Presence rf -> eq_presence (context, lf, rf)
+    and eq_type_args  (context, (lpk, l), (rpk, r))  =
+      (* let lpk = TypeUtils.primary_kind_of_type l in
```
Comment (jamescheney) [inline core/irCheck.ml:437]: Dead code?

File: core/irCheck.ml [hunk 6]
```diff
@@ -473,8 +520,10 @@ struct
       ensure_effect_present_in_row (o#extract_type_equality_context ()) allowed_effects effect_name effect_typ occurrence
 
     method add_typevar_to_context id kind = {< type_var_env = Env.bind id kind type_var_env  >}
-    method remove_typevar_to_context id  = {< type_var_env = Env.unbind id type_var_env >}
+    (* dangerous, unnecessary, and ungrammatical *)
```
Comment (jamescheney) [inline core/irCheck.ml:523]: Dead code?

File: core/irCheck.ml [hunk 7]
```diff
@@ -518,38 +567,42 @@ struct
...
+            *   (fun o quant ->
+            *     let var = Quantifier.to_var quant in
+            *     o#remove_typevar_to_context var) o tyvars in *)
+           let t = Types.for_all (tyvars, t) in
+           TAbs (tyvars, v), t, o
         | TApp (v, ts)  ->
-            let v, t, o = o#value v in
-            let t = Instantiate.apply_type t ts in
-            TApp (v, ts), t, o
+           (* List.iter (TypeUtils.check_type_wellformedness None) ts; *)
```
Comment (jamescheney) [inline core/irCheck.ml:596]: Dead code?

File: core/irCheck.ml [hunk 8]
```diff
@@ -965,17 +1020,17 @@ struct
...
+          let inner_effects_map_from_branches = StringMap.map (fun x -> Present x) branch_presence_spec_types in
           (* We now add all entries from the outer effects that were not touched by the handler to the inner effects *)
           let inner_effects_map = StringMap.fold (fun effect outer_presence_spec map ->
               if StringMap.mem effect inner_effects_map_from_branches then
                 map
               else
                 StringMap.add effect outer_presence_spec map
             )  inner_effects_map_from_branches outer_effects_map in
-          let inner_effects = (inner_effects_map, outer_effects_var, outer_effects_dualized) in
+          let inner_effects = Row (inner_effects_map, outer_effects_var, outer_effects_dualized) in
```
Comment (jamescheney) [inline core/irCheck.ml:1033]: Just curious why `Row` here and not `Effect`?

File: core/irCheck.ml [hunk 9]
```diff
@@ -1262,6 +1328,10 @@ struct
         let tyenv = Env.bind var (info_type info) tyenv in
           (var, info), {< tyenv=tyenv >}
 
+    (* WARNING: use of remove_binder / remove_binding is only sound
```
Comment (jamescheney) [inline core/irCheck.ml:1331]: Should this be an issue/enhancement?

File: core/lib.ml [hunk 10]
```diff
@@ -2,7 +2,6 @@ open CommonTypes
 open List
 
 (*open Value*)
```
Comment (jamescheney) [inline core/lib.ml:4]: Dead code?

File: core/sugartoir.ml [hunk 11]
```diff
@@ -865,6 +867,7 @@ struct
               let qs = List.map SugarQuantifier.get_resolved_exn tyvars in
                 cofv (I.tabstr (qs, v))
           | TAppl (e, tyargs) ->
+(* Debug.print ("TAppl...: "^ Sugartypes.show_phrasenode (TAppl (e, tyargs))); *)
```
Comment (jamescheney) [inline core/sugartoir.ml:870]: Debugging message?

File: core/dune.refactor [hunk 12]
```diff
@@ -0,0 +1,117 @@
+;; TODO remove
```
Comment (jstolarek) [inline core/dune.refactor:1]: I think this file shouldn't be included in the PR.

File: dune [hunk 13]
```diff
@@ -6,14 +6,24 @@
...
+          -g                  ;; Adds debugging information to the resulting executable / library.
+ )))
+(nowarnings
+  (flags (:standard
+          -strict-formats     ;; Disallows legacy formats.
+          -strict-sequence    ;; Enforces the lhs of a sequence to have type `unit'.
+          -safe-string        ;; Immutable strings.
+          -annot              ;; Dumps information per (file) module about types, bindings, tail-calls, etc. Used by some tools.
+          -warn-error -a      ;; Do not treat warnings as errors.
+          -w -A               ;; Ignore all warnings
```
Comment (jstolarek) [inline dune:19]: What's going on here?

Comment (dhil) [reply dune:19]: The build profile `nowarnings` should be deleted unless we elect to keep such a profile. Personally, I don't think we should keep it.

Comment (frank-emrich) [reply dune:19]: I added it for the purposes of the refactoring, but I don't see a point in keeping it

Comment (slindley) [reply dune:19]: I guess it might be useful, but it's pretty easy to add back in if anyone wants to do this again. I'll remove it for now.

File: core/irCheck.ml (hunk 8 above)
Comment (slindley) [reply core/irCheck.ml:1033]: Well we need a row here anyway... really we just need a triple of type `row'`. We couldn't do `Effect ...` on its own but we could do `Effect (Row ...)` which would be even more annoying than what we have now with the `Row` constructor as it would require a further level of boxing/unboxing. That said, it might be worth revisiting this when we rationalise the design further, and perhaps it should morally be `Effect (Row ...)`.

File: core/irCheck.ml (hunk 9 above)
Comment (slindley) [reply core/irCheck.ml:1331]: if you want to make an issue of this then feel free... but really I just left this here for Frank's benefit

File: core/desugarInners.ml [hunk 14]
```diff
@@ -41,9 +41,9 @@ let rec add_extras in_fun tyvars extras tyargs =
              else
                let open CommonTypes.PrimaryKind in
                match SugarQuantifier.get_resolved_exn sq with
-               | _, (Type, sk) -> `Type (Types.fresh_type_variable sk)
-               | _, (Row, sk) -> `Row (Types.make_empty_open_row sk)
-               | _, (Presence, sk) -> `Presence (Types.fresh_presence_variable sk)
+               | _, (Type, sk)     -> Type, Types.fresh_type_variable sk
+               | _, (Row, sk)      -> Row, Types.make_empty_open_row sk
+               | _, (Presence, sk) -> Presence, Types.Present (Types.fresh_presence_variable sk)
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/desugarInners.ml:46]: good catch - I think this is a bug (which apparently isn't exercised by our tests!)

I'll fix it.

File: core/irCheck.ml (hunk 9 above)
Comment (jamescheney) [reply core/irCheck.ml:1331]: Ah, sorry, I was reflexively complaining about commented out stuff.  Never mind.

File: core/generalise.ml [hunk 15]
```diff
@@ -26,118 +26,144 @@ let rec get_type_args : gen_kind -> TypeVarSet.t -> datatype -> type_arg list =
...
+ *       field_vars @ row_vars *)
 
 and get_type_arg_type_args : gen_kind -> TypeVarSet.t -> type_arg -> type_arg list =
-  fun kind bound_vars ->
-    function
-      | `Type t -> get_type_args kind bound_vars t
-      | `Row r -> get_row_type_args kind bound_vars r
-      | `Presence f -> get_presence_type_args kind bound_vars f
+  fun kind bound_vars (pk, t) ->
+  let open PrimaryKind in
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/generalise.ml:141]: my emacs mode doesn't indent here, so for now I'm leaving it as is

File: core/instantiate.ml [hunk 16]
```diff
@@ -36,173 +33,182 @@ type instantiation_maps = (datatype IntMap.t * row IntMap.t * field_spec IntMap.
...
-      StringMap.fold StringMap.add field_env' field_env'', row_var', dual'
+    let field_env'', row_var', dual' = inst_row_var inst_map rec_env row_var dual |> TypeUtils.extract_row_parts in
+    Row (StringMap.fold StringMap.add field_env' field_env'', row_var', dual')
         (* precondition: row_var has been flattened *)
-  and inst_row_var : instantiation_maps -> inst_env -> row_var -> bool -> row = fun inst_maps rec_env row_var dual ->
-    let instr = inst_row inst_maps rec_env in
-    let (_, renv, _) = inst_maps in
-    let (rec_type_env, rec_row_env) = rec_env in
+  and inst_row_var : instantiation_maps -> inst_env -> row_var -> bool -> row = fun inst_map rec_env row_var dual ->
+    (* HACK: fix the ill-formed rows that are introduced in the
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/instantiate.ml:164]: probably

File: core/instantiate.mli (hunk 3 above)
Comment (slindley) [reply core/instantiate.mli:4]: this isn't done and the comment still applies

File: core/irCheck.ml (hunk 6 above)
Comment (slindley) [reply core/irCheck.ml:523]: potentially useful for Frank... so I won't remove it now

File: core/instantiate.ml (hunk 16 above)
[the remark this answers is not in the document]
Comment (slindley) [reply core/instantiate.ml:164]: actually it will become moot in the next phase of types refactoring, so there's probably not much point in creating an issue

File: core/types.ml [hunk 17]
```diff
@@ -1404,93 +1577,100 @@ and unwrap_row : row -> (row * row_var option) = fun (field_env, row_var, dual)
...
-        begin
-          match Unionfind.find point with
-            | `Var _ -> f
-            | `Body f -> normalise_field_spec rec_names f
-        end
-    | `Present t -> `Present (normalise_datatype rec_names t)
-    | _ -> f
+ *)
+and normalise_field_spec rec_names f = normalise_datatype rec_names f
+(* match f with
```
Comment (jamescheney) [inline core/types.ml:1665]: delete dead code?

File: core/types.ml [hunk 18]
```diff
@@ -1507,66 +1687,85 @@ let normalise_row = normalise_row IntSet.empty
 
 (** building quantified types *)
 
+(* let quantifier_of_type_arg =
```
Comment (jamescheney) [inline core/types.ml:1690]: delete dead code?

File: core/types.ml [hunk 19]
```diff
@@ -1598,90 +1797,108 @@ struct
...
-          else
-            (var, (`Recursive, pk_row, `Bound))::(free_bound_row_type_vars (TypeVarSet.add var bound_vars) row)
-      | `Body row -> free_bound_row_type_vars bound_vars row
-  and free_bound_tyarg_vars bound_vars =
-    function
-      | `Type t -> free_bound_type_vars bound_vars t
-      | `Row row -> free_bound_row_type_vars bound_vars row
-      | `Presence f -> free_bound_field_spec_type_vars bound_vars f
+    free_bound_type_vars bound_vars
+      (* | Present t -> free_bound_type_vars bound_vars t
```
Comment (jamescheney) [inline core/types.ml:1868]: remove dead code (x3)?

File: core/types.ml [hunk 20]
```diff
@@ -1692,10 +1909,10 @@ struct
       let bound_vars, vars =
         List.fold_left
           (fun (bound_vars, vars) tyvar ->
-             let var, spec = varspec_of_tyvar tyvar in
-               (TypeVarSet.add var bound_vars, (var, spec)::vars)) (bound_vars, []) tyvars in
-        (bound_vars, List.rev vars) in
-
+            let var, spec = varspec_of_tyvar tyvar in
+             (TypeVarSet.add var bound_vars, (var, spec)::vars)) (bound_vars, []) tyvars in
```
Comment (jamescheney) [inline core/types.ml:1913]: indentation seems off here; putting the `in` at the end on the next line (or unnesting the `let`s a bit) might also help readability

File: core/types.ml [hunk 21]
```diff
@@ -2008,6 +2235,8 @@ struct
        let sd = datatype context p in
 
        let ppr_arrow () =
+         (* Debug.print ("effs: "^show_row effects);
```
Comment (jamescheney) [inline core/types.ml:2238]: delete temporary debugging code?

File: core/types.ml [hunk 22]
```diff
@@ -2396,114 +2647,136 @@ let make_fresh_envs : datatype -> datatype IntMap.t * row IntMap.t * field_spec
...
+  and make_env_f boundvars t = make_env boundvars t
+  and make_env_ta boundvars (pk, t) =
+    let open PrimaryKind in
+    match pk with
+    | Type     -> make_env boundvars t
+    | Row      -> make_env_r boundvars t
+    | Presence -> make_env_f boundvars t in
+  make_env S.empty
+
+  (* and make_env_f boundvars t = make_env boundvars t
```
Comment (jamescheney) [inline core/types.ml:2717]: Dead code?

File: core/types.ml [hunk 23]
```diff
@@ -2697,12 +3001,8 @@ let pp_tycon_spec : Format.formatter -> tycon_spec -> unit = fun fmt t ->
     Format.pp_print_string fmt (string_of_tycon_spec t)
   else
     pp_tycon_spec fmt (decycle_tycon_spec t)
-let pp_row : Format.formatter -> row -> unit = fun fmt t ->
-  if Settings.get print_types_pretty then
-    Format.pp_print_string fmt (string_of_row t)
-  else
-    pp_row fmt (DecycleTypes.row t)
+
```
Comment (jamescheney) [inline core/types.ml:3004]: Unnecessary space?

File: core/unify.ml [hunk 24]
```diff
@@ -82,107 +82,152 @@ inside points *)
...
-      | `Effect l ->
+      | Record l ->
+         begin match unalias t2 with
+         | Record r -> eq_rows (l, r)
+         | _        -> false
+         end
+      | Variant l ->
          begin match unalias t2 with
-         | `Effect r -> eq_rows (l, r)
+         | Variant r -> eq_rows (l, r)
```
Comment (jamescheney) [inline core/unify.ml:141]: I expect this an intentional bugfix?

File: core/unify.ml [hunk 25]
```diff
@@ -82,107 +82,152 @@ inside points *)
...
-and eq_rows : (row * row) -> bool =
-  fun ((lfield_env, lrow_var, ldual), (rfield_env, rrow_var, rdual)) ->
-    eq_field_envs (lfield_env, rfield_env) && eq_row_vars (lrow_var, rrow_var) && ldual=rdual
-and eq_presence =
-  function
-    | `Absent, `Absent -> true
-    | `Present lt, `Present rt -> eq_types (lt, rt)
-    | `Var lpoint, `Var rpoint -> Unionfind.equivalent lpoint rpoint
-    | _, _ -> assert false
+(*       | #session_type as l ->
```
Comment (jamescheney) [inline core/unify.ml:214]: Remove dead code?

File: core/unify.ml [hunk 26]
```diff
@@ -191,16 +236,24 @@ and eq_row_vars (lpoint, rpoint) =
...
-    | `Presence lf, `Presence rf -> eq_presence (lf, rf)
-    | _, _ -> false
+  let open PrimaryKind in
+  fun ((lpk, lt), (rpk, rt)) ->
+  match lpk, rpk with
+  | Type, Type         -> eq_types (lt, rt)
+  | Row, Row           -> eq_types (lt, rt)
+  | Presence, Presence -> eq_presence (lt, rt)
+  | _, _               -> false
+ (* eq_types (l, r) *)
```
Comment (jamescheney) [inline core/unify.ml:251]: Remove dead code?

File: core/unify.ml [hunk 27]
```diff
@@ -1022,24 +1095,30 @@ and unify_rows' : ?var_sk:Subkind.t -> unify_env -> ((row * row) -> unit) =
     unify_both_rigid_with_rec_env rec_env (lrow, rrow) in
 
   let unify_one_rigid ((rigid_field_env, _, _ as rigid_row), (flexible_field_env, _, _ as flexible_row)) =
-    let (rigid_field_env', rigid_row_var', rigid_dual') as rigid_row', rigid_rec_row = unwrap_row rigid_row in
-    let closed = is_closed_row rigid_row' in
-    let (flexible_field_env', flexible_row_var', flexible_dual) as flexible_row', flexible_rec_row = unwrap_row flexible_row in
+    (* FIXME: row = typ was an idiotic mistake! *)
```
Comment (jamescheney) [inline core/unify.ml:1098]: Issue?

File: core/unify.ml [hunk 28]
```diff
@@ -1092,36 +1179,46 @@ and unify_rows' : ?var_sk:Subkind.t -> unify_env -> ((row * row) -> unit) =
...
-     raise (Failure (`Msg ("Couldn't match "^ string_of_type_arg l ^" against "^ string_of_type_arg r)))
+  fun rec_env ((lpk, lt), (rpk, rt)) ->
+  let open PrimaryKind in
+  match lpk, rpk with
+  | Type, Type         -> unify' rec_env (lt, rt)
+  | Row, Row           -> unify' rec_env (lt, rt)
+  | Presence, Presence -> unify_presence' rec_env (lt, rt)
+  | _, _ ->
+     raise (Failure (`Msg ("Couldn't match "^ string_of_type_arg (lpk, lt) ^" against "^ string_of_type_arg (rpk, rt))))
+  (* deferring to unify' means that unify' must handle kinds Row and
```
Comment (jamescheney) [inline core/unify.ml:1213]: remove dead code?

File: core/types.mli [hunk 29]
```diff
@@ -106,38 +77,59 @@ and rec_appl = {
...
+and datatype = typ
+and type_arg = PrimaryKind.t * typ
+and field_spec = typ
+and field_spec_map = field_spec Utility.StringMap.t
+and meta_type_var = typ point
+and meta_row_var = row point
+and meta_presence_var = typ point
+and meta_var = typ point
+and row = typ
+and row' = field_spec_map * row_var * bool
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/types.mli:126]: Yes. We can defer this to part of phase 2.

File: core/typeSugar.ml [hunk 30]
```diff
@@ -3603,35 +3644,36 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * Usage.t =
                           let fieldtype = Types.for_all (pqs, t) in
 
                           (* really we just need to unify the presence
-                             variable with `Presence, but our griper
+                             variable with Presence, but our griper
                              interface doesn't currently support
                              that *)
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/typeSugar.ml:3649]: Hmmm... looking more closely at this unification, it seems completely unnecessary. It just unifies a data type with an unconstrained type variable that is immediately discarded. It will always succeed and serves no purpose. I'll remove this code.

File: core/unify.ml (hunk 24 above)
Comment (slindley) [reply core/unify.ml:141]: Yes.

File: core/unify.ml (hunk 27 above)
Comment (slindley) [reply core/unify.ml:1098]: This will be dealt with in phase 2 of type refactoring - it relates to the `row'` / `row_info` question

File: core/typeSugar.ml [hunk 31]
```diff
@@ -4356,7 +4399,7 @@ and type_binding : context -> binding -> binding * context * Usage.t =
             List.map (WithPos.map ~f:Renamer.rename_recursive_functionnode) defs in
 
           let fresh_tame () = Types.make_empty_open_row default_effect_subkind in
-          (* let fresh_wild () = Types.make_singleton_open_row ("wild", (`Present Types.unit_type)) (lin_any, res_any) in *)
+          (* let fresh_wild () = Types.make_singleton_open_row ("wild", (Present Types.unit_type)) (lin_any, res_any) in *)
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/typeSugar.ml:4402]: it's referenced in commented out code further down, so I'll leave it for now

File: core/typeSugar.ml [hunk 32]
```diff
@@ -4673,9 +4716,10 @@ and type_cp (context : context) = fun {node = p; pos} ->
 
   let unify ~pos ~handle (t, u) = unify_or_raise ~pos:pos ~handle:handle (("<unknown>", t), ("<unknown>", u)) in
 
-  let wild_open = Types.make_singleton_open_row ("wild", `Present Types.unit_type) default_effect_subkind in
-  unify ~pos ~handle:Gripers.cp_wild (`Effect wild_open, `Effect context.effect_row);
+  let wild_open = Types.make_singleton_open_row ("wild", Types.Present Types.unit_type) default_effect_subkind in
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/typeSugar.ml:4719]: Yes, maybe.

File: core/typeUtils.ml [hunk 33]
```diff
@@ -125,67 +134,69 @@ let rec choice_at name t = match concrete_type t with
...
-  | (`Function (`Record row, _, _), _) ->
+  | (ForAll (_, t), true) -> arg_types t
+  | (Function (Record row, _, _), _) ->
       extract_tuple row
-  | (`Lolli (`Record row, _, _), _) ->
+  | (Lolli (Record row, _, _), _) ->
      extract_tuple row
-(*   | `Function (t', _, _) when is_thunk_type t' -> [Types.unit_type] (\* THIS IS A HACK. TODO: Trace down cause of bug. At some point during the compilation process the formal parameter to (() {Op: a -> b} -> c) -> d gets unwrapped yielding a function type composed internally as *)
-(*  `Function ((`Function (), {Op: a -> b}, c) *)
+(*   | Function (t', _, _) when is_thunk_type t' -> [Types.unit_type] (\* THIS IS A HACK. TODO: Trace down cause of bug. At some point during the compilation process the formal parameter to (() {Op: a -> b} -> c) -> d gets unwrapped yielding a function type composed internally as *)
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/typeUtils.ml:177]: I don't know... this is nothing to do with the refactoring, so I'll leave it here for now.

File: core/typevarcheck.ml [hunk 34]
```diff
@@ -124,168 +134,171 @@ let rec is_negative : TypeVarSet.t -> StringSet.t -> int -> datatype -> bool =
...
+and is_negative_lens_type = (*: TypeVarSet.t -> StringSet.t -> int -> Lens.Type.t -> bool =*)
   fun _bound_vars _expanded_apps _var _typ ->
     false
 
 and is_positive : TypeVarSet.t -> StringSet.t -> int -> datatype -> bool =
   fun bound_vars expanded_apps var t ->
     let isp = is_positive bound_vars expanded_apps var in
     let isn = is_negative bound_vars expanded_apps var in
-    let ispr = is_positive_row bound_vars expanded_apps var in
+    let ispr = is_negative_row bound_vars expanded_apps var in
```
[the remark this answers is not in the document]
Comment (slindley) [reply core/typevarcheck.ml:222]: it's a copy-and-paste error - I'll fix it
